File: src/R/tests/testthat/test-synthetic.R
DEFAULT_C <- vimureP$synthetic$DEFAULT_C
DEFAULT_N <- vimureP$synthetic$DEFAULT_N
DEFAULT_M <- vimureP$synthetic$DEFAULT_M
DEFAULT_K <- vimureP$synthetic$DEFAULT_K
DEFAULT_L <- vimureP$synthetic$DEFAULT_L
DEFAULT_AVG_DEGREE <- vimureP$synthetic$DEFAULT_AVG_DEGREE
DEFAULT_ETA <- vimureP$synthetic$DEFAULT_ETA
DEFAULT_C <- vimureP$synthetic$DEFAULT_C
DEFAULT_EXP_IN <- vimureP$synthetic$DEFAULT_EXP_IN
DEFAULT_EXP_OUT <- vimureP$synthetic$DEFAULT_EXP_OUT

skip_if_no_vimure <- function() {
  have_vimure <- reticulate::py_module_available("vimure")
  if (!have_vimure)
    skip("Vimure not available for testing")
}

check_default_values <- function(synth_net){
  expect_equal(synth_net$C, DEFAULT_C)
  expect_equal(synth_net$N, DEFAULT_N)
  expect_equal(synth_net$M, DEFAULT_M)
  expect_equal(synth_net$K, DEFAULT_K)
  expect_equal(synth_net$L, DEFAULT_L)
  expect_equal(synth_net$avg_degree, DEFAULT_AVG_DEGREE)
  expect_equal(unlist(synth_net$Y$shape), c(DEFAULT_L, DEFAULT_N, DEFAULT_N))

  Y <- synth_net$Y$toarray()
  expect_gt(sum(Y), 0)
  expect_lt(max(Y), synth_net$K)
}

check_affinity_matrix <- function(synth_net){
  L <- synth_net$L

  # Check affinity matrix dimensions
  for(i in 1:L){
    current_w <- synth_net$w[i, ,]
    expect_equal(dim(current_w), c(DEFAULT_C, DEFAULT_C))

    diag_w <- diag(current_w)
    non_diag_w <- current_w[as.logical(lower.tri(current_w) + upper.tri(current_w))]
    if(synth_net$structure[i] == "assortative"){
      expect_true(all(sapply(diag_w, function(x) all(x > non_diag_w))))
    } else{
      expect_true(all(sapply(diag_w, function(x) all(x < non_diag_w))))
    }

  }
}

check_Y_change_for_exp_in_ou_change <- function(synth_net, sparse_synth_net){
  another_synth_net <- gm_degree_corrected_sbm(
    N=DEFAULT_N,
    M=DEFAULT_M,
    L=DEFAULT_L,
    K=DEFAULT_K,
    C=DEFAULT_C,
    avg_degree=DEFAULT_AVG_DEGREE,
    exp_in=4,
    exp_out=6,
    sparsify=FALSE
  )

  another_sparse_synth_net <- gm_degree_corrected_sbm(
    N=DEFAULT_N,
    M=DEFAULT_M,
    L=DEFAULT_L,
    K=DEFAULT_K,
    C=DEFAULT_C,
    avg_degree=DEFAULT_AVG_DEGREE,
    exp_in=4,
    exp_out=6,
    sparsify=TRUE
  )

  Y <- synth_net$Y$toarray()
  another_Y <- another_synth_net$Y$toarray()
  sparse_Y <- sparse_synth_net$Y$toarray()
  another_sparse_Y <- another_sparse_synth_net$Y$toarray()

  expect_false(all(Y == another_Y))
  expect_false(all(sparse_Y == another_sparse_Y))
  expect_gt(sum(Y), sum(sparse_Y))
  expect_gt(sum(Y), sum(another_Y))
  expect_gt(sum(another_Y), sum(another_sparse_Y))
}

test_that("Multitensor - Check default values and adjacency matrix structure", {
  skip_if_no_vimure()

  synth_net <- gm_crep(
    N = DEFAULT_N,
    M = DEFAULT_M,
    K = DEFAULT_K,
    L = DEFAULT_L,
    C = DEFAULT_C,
    eta = DEFAULT_ETA,
    avg_degree = DEFAULT_AVG_DEGREE
  )

  check_default_values(synth_net)
  expect_equal(synth_net$eta, DEFAULT_ETA)
  expect_false(is.null(synth_net$ExpM))
})

test_that("gm_standard_sbm - Check default values, adjacency matrix and affinity matrix structure", {
  skip_if_no_vimure()

  synth_net <- gm_standard_sbm(
    N=DEFAULT_N,
    M=DEFAULT_M,
    L=DEFAULT_L,
    K=DEFAULT_K,
    C=DEFAULT_C,
    avg_degree=DEFAULT_AVG_DEGREE,
    sparsify=FALSE,
  )

  check_default_values(synth_net)
  check_affinity_matrix(synth_net)
})

test_that("gm_degree_corrected_sbm - Check default values and adjacency matrix", {
  skip_if_no_vimure()

  synth_net <- gm_degree_corrected_sbm(
    N=DEFAULT_N,
    M=DEFAULT_M,
    L=DEFAULT_L,
    K=DEFAULT_K,
    avg_degree=DEFAULT_AVG_DEGREE,
    C=DEFAULT_C,
    sparsify=FALSE
  )

  sparse_synth_net <- gm_degree_corrected_sbm(
    N=DEFAULT_N,
    M=DEFAULT_M,
    L=DEFAULT_L,
    K=DEFAULT_K,
    C=DEFAULT_C,
    avg_degree=DEFAULT_AVG_DEGREE,
    sparsify=TRUE
  )

  expect_equal(synth_net$exp_in, DEFAULT_EXP_IN)
  expect_equal(synth_net$exp_out, DEFAULT_EXP_OUT)

  expect_equal(length(synth_net$d_in), synth_net$N)
  expect_equal(length(synth_net$d_out), synth_net$N)

  check_Y_change_for_exp_in_ou_change(synth_net, sparse_synth_net)
})

File: src/python/vimure/test/test_synthetic.py
import pytest

import numpy as np
import vimure as vm
import networkx as nx
import sktensor as skt

from vimure.synthetic import (
    DEFAULT_N,
    DEFAULT_M,
    DEFAULT_K,
    DEFAULT_L,
    DEFAULT_C,
    DEFAULT_EXP_IN,
    DEFAULT_EXP_OUT,
    DEFAULT_AVG_DEGREE,
    DEFAULT_SEED,
    DEFAULT_ETA,
)


class TestBaseSyntheticNetwork:
    def check_objects(self, synth_net):
        assert synth_net.C == DEFAULT_C
        assert synth_net.N == DEFAULT_N
        assert synth_net.M == DEFAULT_M
        assert synth_net.L == DEFAULT_L
        assert synth_net.avg_degree == DEFAULT_AVG_DEGREE
        assert synth_net.Y.shape == (DEFAULT_L, DEFAULT_N, DEFAULT_N)

        assert isinstance(synth_net.Y, skt.sptensor) or isinstance(synth_net.Y, skt.dtensor)
        assert synth_net.Y.vals.sum() != 0.0, "Adjacency matrices are empty!"
        assert synth_net.Y.vals.max() < synth_net.K


class TestStandardSBM(TestBaseSyntheticNetwork):
    @classmethod
    def setup_class(cls):
        """
        Creates a single StandardSBM instance object to be shared by tests in this class
        """
        cls.synth_net = vm.synthetic.StandardSBM(
            N=DEFAULT_N,
            M=DEFAULT_M,
            L=DEFAULT_L,
            K=DEFAULT_K,
            C=DEFAULT_C,
            avg_degree=DEFAULT_AVG_DEGREE,
            sparsify=False,
        )

        # TODO: Add specific test cases for networks with overlapping communities
        cls.overlapping_net = vm.synthetic.StandardSBM(
            N=DEFAULT_N,
            M=DEFAULT_M,
            L=DEFAULT_L,
            K=DEFAULT_K,
            C=DEFAULT_C,
            avg_degree=DEFAULT_AVG_DEGREE,
            overlapping=0.1,
            normalization=False,
            corr=0.2,
        )

        cls.sparse_synth_net = vm.synthetic.StandardSBM(
            N=DEFAULT_N,
            M=DEFAULT_M,
            L=DEFAULT_L,
            K=DEFAULT_K,
            C=DEFAULT_C,
            avg_degree=DEFAULT_AVG_DEGREE,
            sparsify=True,
        )

    def test_yields_appropriate_objects(self):
        """
        CHECK matrices and objects have the right types and shapes and values
        """

        self.check_objects(self.synth_net)
        self.check_objects(self.sparse_synth_net)

    def test_yields_appropriate_affinity_matrix(self):
        """
        Focusing on matrix w: did it yield expected values?
        """

        for l in range(self.synth_net.L):

            msg = "Dimmensions of affinity matrix w[l=%d] are incorrect." % l
            assert self.synth_net.w[l].shape == (
                DEFAULT_C,
                DEFAULT_C,
            ), msg

            diag_vals = self.synth_net.w[l].diagonal()

            non_diag_indices = (
                np.ones(shape=self.synth_net.w[l].shape) - np.identity(len(self.synth_net.w[l]))
            ).astype(bool)
            non_diag_vals = self.synth_net.w[l][non_diag_indices]

            msg = "Diagonal values in the affinity matrix should be %s than non-diag."

            # TODO: Remove iteration over C
            for c in enumerate(range(self.synth_net.C)):
                diag_arr, non_diag_arr = np.meshgrid(diag_vals, non_diag_vals)
                if self.synth_net.structure[l] == "assortative":
                    assert np.alltrue(diag_arr > non_diag_arr), msg % "greater"
                else:
                    assert np.alltrue(diag_arr < non_diag_arr), msg % "smaller"


class TestDegreeCorrectedSBM(TestStandardSBM):
    @classmethod
    def setup_class(cls):
        """
        Creates a single StandardSBM instance object to be shared by tests in this class
        """
        cls.synth_net = vm.synthetic.DegreeCorrectedSBM(
            N=DEFAULT_N,
            M=DEFAULT_M,
            L=DEFAULT_L,
            K=DEFAULT_K,
            avg_degree=DEFAULT_AVG_DEGREE,
            C=DEFAULT_C,
            sparsify=False,
        )

        cls.sparse_synth_net = vm.synthetic.DegreeCorrectedSBM(
            N=DEFAULT_N,
            M=DEFAULT_M,
            L=DEFAULT_L,
            K=DEFAULT_K,
            C=DEFAULT_C,
            avg_degree=DEFAULT_AVG_DEGREE,
            sparsify=True,
        )

    def test_yields_appropriate_objects(self):
        """
        CHECK matrices and objects have the right types and shapes and values
        """

        super().test_yields_appropriate_objects()

        assert self.synth_net.exp_in == DEFAULT_EXP_IN
        assert self.synth_net.exp_out == DEFAULT_EXP_OUT

        assert len(self.synth_net.d_in) == self.synth_net.N
        assert len(self.synth_net.d_out) == self.synth_net.N

    def test_degree_correction_was_indeed_applied(self):
        """
        Does the Y matrix also change when we change exp_in and exp_out?
        """
        another_synth_net = vm.synthetic.DegreeCorrectedSBM(
            N=DEFAULT_N,
            M=DEFAULT_M,
            L=DEFAULT_L,
            K=DEFAULT_K,
            C=DEFAULT_C,
            avg_degree=DEFAULT_AVG_DEGREE,
            exp_in=4,
            exp_out=6,
            sparsify=False,
        )
        another_sparse_synth_net = vm.synthetic.DegreeCorrectedSBM(
            N=DEFAULT_N,
            M=DEFAULT_M,
            L=DEFAULT_L,
            K=DEFAULT_K,
            C=DEFAULT_C,
            avg_degree=DEFAULT_AVG_DEGREE,
            exp_in=4,
            exp_out=6,
            sparsify=True,
        )

        np.testing.assert_raises(
            AssertionError,
            np.testing.assert_array_equal,
            self.synth_net.Y.toarray(),
            another_synth_net.Y.toarray(),
        )
        np.testing.assert_raises(
            AssertionError,
            np.testing.assert_array_equal,
            self.sparse_synth_net.Y.toarray(),
            another_sparse_synth_net.Y.toarray(),
        )

        assert np.sum(self.synth_net.Y.vals) > np.sum(
            self.sparse_synth_net.Y.vals
        ), "self.synth_net.Y should be denser than self.sparser_synth_net.Y"
        assert np.sum(self.synth_net.Y.vals) > np.sum(
            another_synth_net.Y.vals
        ), "another_synth_net should be denser than self.synth_net.Y"
        assert np.sum(another_synth_net.Y.vals) > np.sum(
            another_sparse_synth_net.Y.vals
        ), "another_sparser_synth_net.Y should be denser than self.sparser_synth_net.Y"


class TestMultitensor(TestBaseSyntheticNetwork):
    @classmethod
    def setup_class(cls):
        """
        Creates a single Multitensor instance object to be shared by tests in this class
        """
        cls.synth_net = vm.synthetic.Multitensor(
            N=DEFAULT_N,
            M=DEFAULT_M,
            K=DEFAULT_K,
            L=DEFAULT_L,
            C=DEFAULT_C,
            eta=DEFAULT_ETA,
            avg_degree=DEFAULT_AVG_DEGREE,
        )

    def test_yields_appropriate_objects(self):
        """
        CHECK matrices and objects have the right types and shapes and values
        """

        self.check_objects(self.synth_net)

        assert self.synth_net.eta == DEFAULT_ETA
        assert self.synth_net.ExpM is not None
